src/pybind/ceph_argparse.py

```diff
@@ -168,32 +168,34 @@ def complete(self, s):
         return []
 
     @staticmethod
-    def _compound_type_to_argdesc(tp, attrs):
+    def _compound_type_to_argdesc(tp, attrs, has_default, positional):
```

Comment:
we have following call graph

- load_func_metadata(func)
  - to_argdesc(tp, attrs, has_default) // set `attrs['req']` if `has_default`
    - tp.argdesc(attrs)
    - _compound_type_to_argdesc(tp, attrs)
      - to_argdesc(tp.inner_tp, attrs, has_default=True)  # tp is non-Union
      - to_argdesc(tp.inner_tp, attrs, has_default=False) # tp Union

why shall we populate `has_default` down to `_compound_type_to_argdesc()` even if we've already set `attr['req']` in the caller?

Reply:
yep, i wasn't reading the code carefully!